pkg/controller/podautoscaler/horizontal.go

```diff
@@ -139,141 +150,140 @@ func (a *HorizontalController) Run(stopCh <-chan struct{}) {
-	if err != nil {
-		lastScaleTime := getLastScaleTime(hpa)
-		if time.Now().After(lastScaleTime.Add(upscaleForbiddenWindow)) {
-			a.eventRecorder.Event(hpa, v1.EventTypeWarning, "FailedGetMetrics", err.Error())
-		} else {
-			a.eventRecorder.Event(hpa, v1.EventTypeNormal, "MetricsNotAvailableYet", err.Error())
-		}
-
-		return 0, nil, time.Time{}, fmt.Errorf("failed to get CPU utilization: %v", err)
-	}
-
-	if desiredReplicas != currentReplicas {
-		a.eventRecorder.Eventf(hpa, v1.EventTypeNormal, "DesiredReplicasComputed",
-			"Computed the desired num of replicas: %d (avgCPUutil: %d, current replicas: %d)",
-			desiredReplicas, utilization, scale.Status.Replicas)
-	}
-
-	return desiredReplicas, &utilization, timestamp, nil
-}
-
-// computeReplicasForCustomMetrics computes the desired number of replicas based on the CustomMetrics passed in cmAnnotation
-// as json-serialized extensions.CustomMetricsTargetList.
-// Returns number of replicas, metric which required highest number of replicas,
-// status string (also json-serialized extensions.CustomMetricsCurrentStatusList),
-// last timestamp of the metrics involved in computations or error, if occurred.
-func (a *HorizontalController) computeReplicasForCustomMetrics(hpa *autoscaling.HorizontalPodAutoscaler, scale *extensionsv1beta1.Scale,
-	cmAnnotation string) (replicas int32, metric string, status string, timestamp time.Time, err error) {
-
-	if cmAnnotation == "" {
-		return
-	}
+// Computes the desired number of replicas for the metric specifications listed in the HPA, returning the maximum
+// of the computed replica counts, a description of the associated metric, and the statuses of all metrics
+// computed.
+func (a *HorizontalController) computeReplicasForMetrics(hpa *autoscalingv2.HorizontalPodAutoscaler, scale *extensions.Scale,
+	metricSpecs []autoscalingv2.MetricSpec) (replicas int32, metric string, statuses []autoscalingv2.MetricStatus, timestamp time.Time, err error) {
 
 	currentReplicas := scale.Status.Replicas
 
-	var targetList extensionsv1beta1.CustomMetricTargetList
-	if err := json.Unmarshal([]byte(cmAnnotation), &targetList); err != nil {
-		a.eventRecorder.Event(hpa, v1.EventTypeWarning, "FailedParseCustomMetricsAnnotation", err.Error())
-		return 0, "", "", time.Time{}, fmt.Errorf("failed to parse custom metrics annotation: %v", err)
-	}
-	if len(targetList.Items) == 0 {
-		a.eventRecorder.Event(hpa, v1.EventTypeWarning, "NoCustomMetricsInAnnotation", err.Error())
-		return 0, "", "", time.Time{}, fmt.Errorf("no custom metrics in annotation")
-	}
-
-	statusList := extensionsv1beta1.CustomMetricCurrentStatusList{
-		Items: make([]extensionsv1beta1.CustomMetricCurrentStatus, 0),
-	}
+	statuses = make([]autoscalingv2.MetricStatus, len(metricSpecs))
 
-	for _, customMetricTarget := range targetList.Items {
+	for i, metricSpec := range metricSpecs {
 		if scale.Status.Selector == nil {
```

Comment:
I don't understand this check: scale.Status.Selector is array, we should use len to get its length. I guess this check will be always false.